### ISSUE TITLE
Enable collection of more game data when simulating

### ISSUE THREAD

**Issue by danmkent** [nicolodavis/boardgame.io#494]
There are lots of analysis options that would be opened up if it was possible to collect more information about what happens during games when using Simulate.

For example, if the sequence of moves could be accessed, it would mean that we could see how often different kinds of moves are used. If mid-game states could be accessed then things like average scores at different points in the game could be looked at.  If the enumerated moves could be accessed then the average branching factor could be calculated.

Rather than picking specific things and making them part of the game state, I propose enabling code that uses Simulate to pass in handler functions that will be called as the simulation reaches particular events (entering and leaving stages, turns and phases I guess) and that the Bot enables a handler function to be specified that will be called when a bot has enumerated its moves for a turn. 

With this approach, it should be possible for code to grab what it needs without adding to the overall state overhead.

**Comment by nicolodavis** [comment]
Just collecting the ideas we have so far:

#### Default stats that we export if no configuration is passed

```js
const stats = Simulate(game);
```

1. Minimum, maximum and average number of turns per phase / game.
2. Minimum, maximum and average number of moves per turn.
3. Counters for each move played.
4. Same numbers, but reported per player.

#### Customization

Maybe let's have a `stats` object that's passed around that can be intercepted at various points in the game in order to add stuff to it (it's really like `G`, except that you maintain statistics rather than game state).  This part follows a very similar API to that used to create the actual game logic:

```js
const stats = Simulate(game, {
  turn: {
    onBegin: (stats, { G, ctx }) => {
      return { ...stats, /* add stuff to stats */ };
    },
    onEnd:
    onMove:
  },

  phase: {
    onBegin:
    onEnd:
  },
});
```

#### Conditions

Should we have triggers when certain conditions are met?

```js
const states = Simulate(game, {
  triggers: {
     // increments a counter whenever this returns true at the end of a move or event.
    'trigger-1': (G, ctx) => bool,

    // long form that allows you to do arbitrary things to `stats` when a condition is met.
    'trigger-2': {
      condition: (G, ctx) => bool,
      reduce: (stats) => { ... }
    }
  }
});
```

**Comment by danmkent** [comment]
The ability to inject an arbitrary function to add things to the stats object that will be returned would be perfect. 

I like the idea of mirroring the API of the game object with the same set of lifecycle events to hook into.

I probably wouldn't use the trigger API (I'd want to be explicit about when I am collecting stats) but I can imagine people preferring that.

I think the return object from Simulate should include the final game state `{G, ctx}` along with the stats object.

I think there is an argument for returning the complete list of moves taken in the game in a separate property of the object returned by Simulate. If the RNG seed used for the game is also included, that would enable game runs generated by Simulate to be replayed. (This could also be nice for providing a way for manually played games to be stored and replayed later)

For the default stats collected, I guess it's a case of finding the right balance between collecting things that most people will want vs adding overhead for things that are more specific. I will probably want more than most people will be after (e.g. actual distributions for most values rather than just maximum, minimum and mean) but I wouldn't want to introduce too much overhead for those who don't need that level of detail.

I think a minimal set of default stats and then a good set of recipes for collecting other things would be the best approach. Eventually, maybe the plugin system could be used to provide things that are reusable but not universally needed.

**Comment by nicolodavis** [comment]
> I think the return object from Simulate should include the final game state {G, ctx} along with the stats object.

Yep, I agree that we should return an object that contains the stats, game state and other stuff.

I also like the idea of using a minimal set of stats as the default.
